pkg/proxy/util/service.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+)
+
+// ProxyServiceMap maps a service to its service info.
+type ProxyServiceMap map[proxy.ServicePortName]proxy.ServiceInfo
+
+// endpointsChange stores both the previous and current ProxyServiceMap.
+type serviceChange struct {
+	// previous ProxyServiceMap.
+	previous ProxyServiceMap
+	// current ProxyServiceMap.
+	current ProxyServiceMap
+}
+
+// ServiceChangeMap maps a service to its serviceChange.
+type ServiceChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items is a service to its serviceChange map.
+	items map[types.NamespacedName]*serviceChange
+}
+
+func NewServiceChangeMap() ServiceChangeMap {
+	return ServiceChangeMap{
+		items: make(map[types.NamespacedName]*serviceChange),
+	}
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it.
+// It returns all other ProxyEndpointsMap's service ports.
+func (sm *ProxyServiceMap) Merge(other ProxyServiceMap) sets.String {
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		existingPorts.Insert(svcPortName.Port)
+		_, exists := (*sm)[svcPortName]
+		if !exists {
+			glog.V(1).Infof("Adding new service port %q at %s:%d/%s", svcPortName, info.ClusterIP(), info.Port(), info.Protocol())
+		} else {
+			glog.V(1).Infof("Updating existing service port %q at %s:%d/%s", svcPortName, info.ClusterIP(), info.Port(), info.Protocol())
+		}
+		(*sm)[svcPortName] = info
+	}
+	return existingPorts
+}
+
+// Unmerge deletes current ProxyServiceMap's elements which are in other ProxyEndpointsMap.
+func (sm *ProxyServiceMap) Unmerge(other ProxyServiceMap, existingPorts, staleServices sets.String) {
+	for svcPortName := range other {
+		if existingPorts.Has(svcPortName.Port) {
```

Comment:
I feel like existingPorts is redundant.  This is a set of port nums/names for a give Service (see above feedback - this unstated assumption is not clear).  It's not clear if that is the ports on `sm` or on `other`, but in either case we could iterate here and eliminate the argument?

Reply:
Nice finding!

I agree, `existingPorts` is redundant and never used in this function. We can safely eliminate  it.

Reply:
It's fixed now. PTAL.

Reply:
> I feel like existingPorts is redundant. ...but in either case we could iterate here and eliminate the argument?

I took another look and found `existingPorts` is not redundant.

Reply:
`existingPorts` is used to store all the unique newly merged services' identifier and help us tell if a service is updated or deleted. I already added the comment message, PTAL.